### PR TITLE
[dataset] update dataset `SendMgmtGet` APIs and add support in spinel/NCP

### PR DESCRIPTION
**Pull request by abtink**
This PR contains two related commits:

**[dataset] update dataset `SendMgmtGet` APIs and implementation**
    
This commit updates the APIs `otDatasetSendMgmtActiveGet` and
`otDatasetSendMgmtPendingGet` to be similar to MgmtSet APIs and
use an `otOperationalDataset` as input to provide the list of TLVs
to request. The implementation in `DatasetManager` and their use
in in CLI is also updated.

**[ncp] add properties for sending MGMT_GET for Active/Pending Dataset**
    
This commit defines new spinel properties `MGMT_GET_ACTIVE_DATASET`
and `THREAD_MGMT_GET_PENDING_DATASET` (with their set handlers) to add
support for sending `MGMT_GET` meshcop command for Active/Pending
Operational Dataset. It also renames the existing properties used for
sending `MGMT_SET` (include `MGMT_SET` in the property name).
    
It also adds a new property to allow an optional destination IPv6
address to be specified (for sending `MGMT_GET` command) as part
of spinel dictionary representation of Dataset instance.

-----

Related PR in wpantund: https://github.com/openthread/wpantund/pull/327

Edit: I combined the two PRs into one now.

### CODE REVIEW

File: include/openthread/dataset_ftd.h [hunk 1]
```diff
@@ -77,19 +77,24 @@ OTAPI otError OTCALL otDatasetSetPending(otInstance *aInstance, const otOperatio
 /**
  * This function sends MGMT_ACTIVE_GET.
  *
+ * The `mIs<Field>Set` entries in @p aDataset are used to derive the list of TLVs to be requested. The actual
+ * values associated with the fields in @p aDataset are ignored.
+ *
  * @param[in]  aInstance  A pointer to an OpenThread instance.
- * @param[in]  aTlvTypes  A pointer to the TLV Types.
- * @param[in]  aLength    The length of TLV Types.
+ * @param[in]  aDataset   A pointer to an Operational Dataset.
```
[the remark this answers is not in the document]
Comment (abtink) [reply include/openthread/dataset_ftd.h]: All other public dataset APIs use abstracted type `otOperationalDataset` which basically hides the TLV encoding.
Users of OT (e.g., wpantund) may not have any notion of the TLV types.  

This change makes the `otDatasetSendMgmtActiveGet()` be similar/same paramters) as in the related API `otDatasetSendMgmtActiveSet()`. That is, both APIs work with `otOperationalDataset` abstracted types for dataset related fields yet provide a way to define additional raw TLVs to be requested (if needed).

File: include/openthread/dataset_ftd.h [hunk 2]
```diff
@@ -77,19 +77,24 @@ OTAPI otError OTCALL otDatasetSetPending(otInstance *aInstance, const otOperatio
...
+ * @param[in]  aTlvs      A pointer to any additional raw TLVs to include.
```
[the remark this answers is not in the document]
Comment (abtink) [reply include/openthread/dataset_ftd.h]: Any additional raw TLVs to request, similar to `otDatasetSendMgmtActiveSet()` API.

[the remark this answers is not in the document]
Comment (abtink) [reply include/openthread/dataset_ftd.h]: Will rename this to `aTlvTypes` and update the comment to clarify this is an array of TLV types to be requested (when sending MGMT_GET).

File: include/openthread/dataset_ftd.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (librasungirl) [reply include/openthread/dataset_ftd.h]: For `Get()`, User only need specify a list of types, different from `Set()` which need values  either with `otOperationalDataset` which has the `(flag, value)`s or with `aTlvs` which has complete `(type, len, value)`s;  
And for `Get()`, passing `otOperationalDataset` has has the flag named as `mIsXXXSet` may bring some confusion, though noted in the doxygen; enumerating types in public header files instead might be enough to let User know the notion of types.

Thoughts?

[the remark this answers is not in the document]
Comment (bukepo) [reply include/openthread/dataset_ftd.h]: The new Get API seems to be less feasible and complicated, although it's more similar to the corresponding Set API. User not only should know definition of `aTlvTypes`, but also the data structure of `otOperationalDataset`.

[the remark this answers is not in the document]
Comment (abtink) [reply include/openthread/dataset_ftd.h]: In the new pushed commit I changed the API to use a new struct type `otOperationalDatasetComponents` which is bit-filed flags specifying which components are present in an Active/Pending Dataset. The `otOperationalDataset` is updated to include the same structure.  This should help with confusion that `MgmtGet` would not use/need the values and needs a set of flags.

This should give us full flexibility while still being consistent with other existing APIs.

File: examples/drivers/windows/otLwf/iocontrol.c [hunk 3]
```diff
@@ -5918,20 +5918,22 @@ otLwfIoCtl_otSendActiveGet(
     *OutBufferLength = 0;
     UNREFERENCED_PARAMETER(OutBuffer);
 
-    if (InBufferLength >= sizeof(uint8_t))
+    if (InBufferLength >= sizeof(otOperationalDatasetComponents) + sizeof(uint8_t))
     {
+        const otOperationalDatasetComponents *aDatasetComp = (otOperationalDatasetComponents*)InBuffer;
         uint8_t aLength = *(uint8_t*)InBuffer;
         PUCHAR aTlvTypes = aLength == 0 ? NULL : InBuffer + sizeof(uint8_t);
```
Comment (librasungirl) [inline examples/drivers/windows/otLwf/iocontrol.c]: should here update above two lines as below?
```
uint8_t aLength = *(uint8_t*)(InBuffer + sizeof(otOperationalDatasetComponents))
PUCHAR aTlvTypes = aLength == 0 ? NULL : InBuffer + sizeof(otOperationalDatasetComponents) + sizeof(uint8_t);
```

File: examples/drivers/windows/otLwf/iocontrol.c [hunk 4]
```diff
@@ -5997,20 +5999,22 @@ otLwfIoCtl_otSendPendingGet(
     *OutBufferLength = 0;
     UNREFERENCED_PARAMETER(OutBuffer);
 
-    if (InBufferLength >= sizeof(uint8_t))
+    if (InBufferLength >= sizeof(otOperationalDataset) + sizeof(uint8_t))
     {
+        const otOperationalDatasetComponents *aDatasetComp = (otOperationalDatasetComponents*)InBuffer;
         uint8_t aLength = *(uint8_t*)InBuffer;
         PUCHAR aTlvTypes = aLength == 0 ? NULL : InBuffer + sizeof(uint8_t);
```
Comment (librasungirl) [inline examples/drivers/windows/otLwf/iocontrol.c]: update `aLength` and `aTlvTypes` as above comment as well?

File: include/openthread/dataset_ftd.h [hunk 5]
```diff
@@ -111,19 +113,24 @@ OTAPI otError OTCALL otDatasetSendMgmtActiveSet(otInstance *                aIns
 /**
  * This function sends MGMT_PENDING_GET.
  *
- * @param[in]  aInstance  A pointer to an OpenThread instance.
- * @param[in]  aTlvTypes  A pointer to the TLV Types.
- * @param[in]  aLength    The length of TLV Types.
- * @param[in]  aAddress   A pointer to the IPv6 destination, if it is NULL, will use Leader ALOC as default.
+ * The `mIs<Field>Set` entries in @p aDataset are used to derive the list of TLVs to be requested. The actual
```
Comment (librasungirl) [inline include/openthread/dataset_ftd.h]: update the comments accordingly here?

Comment (librasungirl) [inline include/openthread/dataset_ftd.h]: update the comments accordingly here?

File: src/ncp/spinel.h [hunk 6]
```diff
@@ -1408,6 +1416,52 @@ typedef enum {
...
+     * The formatting of this property follows the same rules as in SPINEL_PROP_THREAD_MGMT_GET_ACTIVE_DATASET.
+     *
+     * This is write-only property. When written, it triggers a MGMT_PENDING_GET meshcop command to be sent to leader
+     * with the given Dataset. The spinel frame response should be a `LAST_STATUS` with the status of the transmission
+     * of MGMT_PENDING_GET command.
+     *
+     */
+    SPINEL_PROP_THREAD_MGMT_GET_PENDING_DATASET = SPINEL_PROP_THREAD_EXT__BEGIN + 38,
+
+    /// Operational Dataset (MGM_GET) Destination IPv6 Address
```
Comment (librasungirl) [inline src/ncp/spinel.h]: typo: `MGM_GET`->`MGMT_GET`

File: examples/drivers/windows/otLwf/iocontrol.c (hunk 3 above)
Comment (abtink) [reply examples/drivers/windows/otLwf/iocontrol.c]: Yes. Thanks. Changed it.

File: doc/spinel-protocol-src/spinel-tech-thread.md [hunk 7]
```diff
@@ -583,3 +587,61 @@ Data per item is:
 * `6` : Target IPv6 address
 * `S` : RLOC16 of target
 * `C` : Age (order of use, 0 indicates most recently used entry)
+
+### PROP 5412: SPINEL_PROP_THREAD_MGMT_GET_ACTIVE_DATASET (#prop-thread-mgmt-get-active-dataset)
```
Comment (librasungirl) [inline doc/spinel-protocol-src/spinel-tech-thread.md]: Here should be `PROP 5413`

File: doc/spinel-protocol-src/spinel-tech-thread.md [hunk 8]
```diff
@@ -583,3 +587,61 @@ Data per item is:
...
+should be a `LAST_STATUS` with the status of the transmission
+of MGMT_ACTIVE_GET command.
+
+In addition to supported properties in
+SPINEL_PROP_THREAD_MGMT_SET_ACTIVE_DATASET, the following property
+can be optionally included in the Dataset:
+
+*    SPINEL_PROP_DATASET_DEST_ADDRESS
+
+### PROP 5413: SPINEL_PROP_THREAD_MGMT_GET_PENDING_DATASET (#prop-thread-mgmt-get-pending-dataset)
```
Comment (librasungirl) [inline doc/spinel-protocol-src/spinel-tech-thread.md]: Here should be `PROP 5414`

File: doc/spinel-protocol-src/spinel-tech-thread.md [hunk 9]
```diff
@@ -583,3 +587,61 @@ Data per item is:
...
+`LAST_STATUS` with the status of the transmission of
+MGMT_PENDING_GET command.
+
+In addition to supported properties in
+SPINEL_PROP_THREAD_MGMT_SET_PENDING_DATASET, the following property
+can be optionally included the Dataset:
+
+*    SPINEL_PROP_DATASET_DEST_ADDRESS
+
+### PROP 5414: SPINEL_PROP_DATASET_DEST_ADDRESS (#prop-dataset-dest-address)
```
Comment (librasungirl) [inline doc/spinel-protocol-src/spinel-tech-thread.md]: Here should be `PROP 5415`

File: doc/spinel-protocol-src/spinel-tech-thread.md (hunk 7 above)
Comment (abtink) [reply doc/spinel-protocol-src/spinel-tech-thread.md]: Indeed. Thanks. Will rebase on master and fix.

File: src/core/api/dataset_ftd_api.cpp [hunk 10]
```diff
@@ -87,14 +89,15 @@ otError otDatasetSendMgmtActiveSet(otInstance *                aInstance,
     return instance.GetThreadNetif().GetActiveDataset().SendSetRequest(*aDataset, aTlvs, aLength);
 }
 
-otError otDatasetSendMgmtPendingGet(otInstance *        aInstance,
-                                    const uint8_t *     aTlvTypes,
-                                    uint8_t             aLength,
-                                    const otIp6Address *aAddress)
+otError otDatasetSendMgmtPendingGet(otInstance *                          aInstance,
+                                    const otOperationalDatasetComponents *aDatasetComponents,
+                                    const uint8_t *                       aTlvs,
```
Comment (jwhui) [inline src/core/api/dataset_ftd_api.cpp]: nit: `aTlvs` -> `aTlvTypes`

Comment (abtink) [reply src/core/api/dataset_ftd_api.cpp]: @jwhui coming to change this.... looks you already fix this... Thanks :)

**Pull request merged**